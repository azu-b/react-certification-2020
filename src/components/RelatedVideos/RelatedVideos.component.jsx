import React from 'react';
import { Link } from 'react-router-dom';
import {
  Container,
  Item,
  Thumbnail,
  Title,
  RelatedVideosTitle,
  VideosContainer,
  Error,
} from './RelatedVideos.styled';

const RelatedVideos = (props) => {
  const { videos } = props;

  return videos ? (
    <Container>
      <RelatedVideosTitle>Related Videos</RelatedVideosTitle>
      <VideosContainer data-testid="video-list">
        {videos.map((video) => {
          const { id, title, thumbnails } = video;
          const { default: defaultThumbnail } = thumbnails;
          const { url } = defaultThumbnail;

          return (
            <Link data-testid="link" key={id} to={`/${id}`}>
              <Item data-testid="video-item">
                <Thumbnail data-testid="thumbnail" src={url} />
                <Title data-testid="title">{title}</Title>
              </Item>
            </Link>
          );
        })}
      </VideosContainer>
    </Container>
  ) : (
    <Error>No related videos were provided</Error>
  );
};

export default RelatedVideos;
