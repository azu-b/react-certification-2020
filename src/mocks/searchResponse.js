const searchResponse = {
  kind: 'youtube#searchListResponse',
  etag: 'n9Y42Gm30xxTdZ6Bc6y8YPUX454',
  nextPageToken: 'CAUQAA',
  regionCode: 'MX',
  pageInfo: {
    totalResults: 1000000,
    resultsPerPage: 5,
  },
  items: [
    {
      kind: 'youtube#searchResult',
      etag: 'Jt8LbUC1T4vFrfk8d51TBvmS9B0',
      id: {
        kind: 'youtube#video',
        videoId: '3w5hjn7sY8g',
      },
      snippet: {
        publishedAt: '2020-10-05T00:00:00Z',
        channelId: 'UCq8up0Ew9K0IiRxbz75BZsg',
        title: 'HIGH ELO Best Champions TIER List - League of Legends Patch 10.20',
        description:
          'Want instant, easily-accessible, 24-7 coaching from high elo players? Then check out our website: https://www.proguides.com/yt Patch 10.20 comes with ...',
        thumbnails: {
          default: {
            url: 'https://i.ytimg.com/vi/3w5hjn7sY8g/default.jpg',
            width: 120,
            height: 90,
          },
          medium: {
            url: 'https://i.ytimg.com/vi/3w5hjn7sY8g/mqdefault.jpg',
            width: 320,
            height: 180,
          },
          high: {
            url: 'https://i.ytimg.com/vi/3w5hjn7sY8g/hqdefault.jpg',
            width: 480,
            height: 360,
          },
        },
        channelTitle: 'ProGuides Challenger League of Legends Guides',
        liveBroadcastContent: 'none',
        publishTime: '2020-10-05T00:00:00Z',
      },
    },
    {
      kind: 'youtube#searchResult',
      etag: 'e0RIH0Ar2o0ZTxHcaq70cavX5Xo',
      id: {
        kind: 'youtube#video',
        videoId: 'RR0JFDWuceo',
      },
      snippet: {
        publishedAt: '2020-10-04T09:00:08Z',
        channelId: 'UClh5azhOaKzdlThQFgoq-tw',
        title:
          'When LOL Players PERFECT ESCAPE... 200 IQ ESCAPES MONTAGE (League of Legends)',
        description:
          'When LOL Players PERFECT ESCAPE... Best Pro Player Escapes, Perfect Timing Moments, 200 IQ Tricks and Jukes in a League of Legends Montage ...',
        thumbnails: {
          default: {
            url: 'https://i.ytimg.com/vi/RR0JFDWuceo/default.jpg',
            width: 120,
            height: 90,
          },
          medium: {
            url: 'https://i.ytimg.com/vi/RR0JFDWuceo/mqdefault.jpg',
            width: 320,
            height: 180,
          },
          high: {
            url: 'https://i.ytimg.com/vi/RR0JFDWuceo/hqdefault.jpg',
            width: 480,
            height: 360,
          },
        },
        channelTitle: 'Protatomonster',
        liveBroadcastContent: 'none',
        publishTime: '2020-10-04T09:00:08Z',
      },
    },
    {
      kind: 'youtube#searchResult',
      etag: '6VVytJOelo1z-FdT6KmNHjfE0AY',
      id: {
        kind: 'youtube#video',
        videoId: 'o1YfnQ5enz8',
      },
      snippet: {
        publishedAt: '2020-10-05T00:03:51Z',
        channelId: 'UCv7PqO2lW3eK16sQngYPwsA',
        title:
          'THIS ITEM MAKE SHACO PERMINENTLY INVISIBLE (NO COUNTER) - BunnyFuFuu | League Of Legends | Season 11',
        description:
          'Shaco is always invisible you gotta check this out $2500 CASH or GAMING SETUP WORLDWIDE GIVEAWAY LINK ...',
        thumbnails: {
          default: {
            url: 'https://i.ytimg.com/vi/o1YfnQ5enz8/default.jpg',
            width: 120,
            height: 90,
          },
          medium: {
            url: 'https://i.ytimg.com/vi/o1YfnQ5enz8/mqdefault.jpg',
            width: 320,
            height: 180,
          },
          high: {
            url: 'https://i.ytimg.com/vi/o1YfnQ5enz8/hqdefault.jpg',
            width: 480,
            height: 360,
          },
        },
        channelTitle: 'BunnyFuFuu',
        liveBroadcastContent: 'none',
        publishTime: '2020-10-05T00:03:51Z',
      },
    },
    {
      kind: 'youtube#searchResult',
      etag: 'muLyNyvSkC63p1I96G_nMt62Sp4',
      id: {
        kind: 'youtube#video',
        videoId: 'aR-KAldshAE',
      },
      snippet: {
        publishedAt: '2020-01-09T12:00:16Z',
        channelId: 'UC2t5bjwHdUX4vM2g8TRDq5g',
        title:
          'Warriors | Season 2020 Cinematic - League of Legends (ft. 2WEI and Edda Hayes)',
        description:
          'Warriors: We are warriors. Season 2020 has begun. LISTEN NOW: https://smarturl.it/warriorslol Spotify: http://smarturl.it/warriorslol/spotify Apple Music: ...',
        thumbnails: {
          default: {
            url: 'https://i.ytimg.com/vi/aR-KAldshAE/default.jpg',
            width: 120,
            height: 90,
          },
          medium: {
            url: 'https://i.ytimg.com/vi/aR-KAldshAE/mqdefault.jpg',
            width: 320,
            height: 180,
          },
          high: {
            url: 'https://i.ytimg.com/vi/aR-KAldshAE/hqdefault.jpg',
            width: 480,
            height: 360,
          },
        },
        channelTitle: 'League of Legends',
        liveBroadcastContent: 'none',
        publishTime: '2020-01-09T12:00:16Z',
      },
    },
    {
      kind: 'youtube#searchResult',
      etag: 'dJDFTqE2xU-_WNKHcVBC4NkbT4g',
      id: {
        kind: 'youtube#video',
        videoId: 'pngf7xqrsDU',
      },
      snippet: {
        publishedAt: '2020-10-04T12:00:10Z',
        channelId: 'UCq8up0Ew9K0IiRxbz75BZsg',
        title:
          'NEW UPDATES: Seraphine Abilities LEAKED? + Upcoming JG &amp; Mid Nerfs &amp; MORE - League of Legends',
        description:
          'Want instant, easily-accessible, 24-7 coaching from high elo players? Then check out our website: https://www.proguides.com/yt Come join our Discord server ...',
        thumbnails: {
          default: {
            url: 'https://i.ytimg.com/vi/pngf7xqrsDU/default.jpg',
            width: 120,
            height: 90,
          },
          medium: {
            url: 'https://i.ytimg.com/vi/pngf7xqrsDU/mqdefault.jpg',
            width: 320,
            height: 180,
          },
          high: {
            url: 'https://i.ytimg.com/vi/pngf7xqrsDU/hqdefault.jpg',
            width: 480,
            height: 360,
          },
        },
        channelTitle: 'ProGuides Challenger League of Legends Guides',
        liveBroadcastContent: 'none',
        publishTime: '2020-10-04T12:00:10Z',
      },
    },
  ],
};

export default searchResponse;
