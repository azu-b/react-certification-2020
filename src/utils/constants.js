const AUTH_STORAGE_KEY = 'wa_cert_authenticated';
const USER_STORAGE_KEY = 'wa_cert_auth_user';

export { AUTH_STORAGE_KEY, USER_STORAGE_KEY };
