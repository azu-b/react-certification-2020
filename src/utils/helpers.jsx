import React from 'react';
import { Router, Route } from 'react-router-dom';
import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';

const renderWithRouterMatch = (
  ui,
  {
    path = '/',
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {}
) => {
  return {
    ...render(
      <Router history={history}>
        <Route path={path} component={ui} />
      </Router>
    ),
  };
};

const cleanYouTubeResponse = (videos) => {
  const cleanedVideos = videos.map((video) => {
    return {
      id: video?.id?.videoId || '',
      title: video?.snippet?.title || '',
      description: video?.snippet?.description || '',
      thumbnails: video?.snippet?.thumbnails || '',
    };
  });

  return cleanedVideos;
};

const storage = {
  get(key) {
    try {
      const rawValue = window.localStorage.getItem(key);
      return JSON.parse(rawValue);
    } catch (error) {
      console.error(`Error parsing storage item "${key}".`);
      return null;
    }
  },

  set(key, value) {
    window.localStorage.setItem(key, JSON.stringify(value));
  },
};

export { renderWithRouterMatch, cleanYouTubeResponse, storage };
