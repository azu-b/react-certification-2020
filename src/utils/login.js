const mockedUser = {
  id: '95',
  name: 'Rayo McQueen',
  user: 'RayoMcQueen',
};

const LogInAPI = async (username, password) => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      if (username === 'RayoMcQueen' && password === 'Kuchao!') {
        return resolve(mockedUser);
      }
      return reject(new Error('Username or password invalid'));
    }, 500);
  });
};

export default LogInAPI;
