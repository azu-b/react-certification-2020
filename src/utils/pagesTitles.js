const appTitle = "Azu's React Cert Challenge";

const homeTitle = 'Home';

const loginTitle = 'Log In';

const favoritesTitle = 'Favorite Videos';

const buildTitle = (pageTitle) => {
  const title = `${pageTitle} | ${appTitle}`;
  return title;
};

export { homeTitle, loginTitle, favoritesTitle, buildTitle };
