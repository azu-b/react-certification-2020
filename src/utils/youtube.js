import axios from 'axios';
import YOUTUBE_API_KEY from './env';

const YouTubeAPI = axios.create({
  baseURL: 'https://www.googleapis.com/youtube/v3/',
  params: {
    part: 'id,snippet',
    key: YOUTUBE_API_KEY,
  },
});

const cleanYouTubeResponse = (videos) => {
  const cleanedVideos = videos.map((video) => {
    return {
      id: video?.id?.videoId || '',
      title: video?.snippet?.title || '',
      description: video?.snippet?.description || '',
      thumbnails: video?.snippet?.thumbnails || '',
    };
  });

  return cleanedVideos;
};

export { YouTubeAPI, cleanYouTubeResponse };
