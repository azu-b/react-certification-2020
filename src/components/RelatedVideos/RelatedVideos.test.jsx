import React from 'react';
import { render, cleanup, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import pretty from 'pretty';
import RelatedVideos from './index';
import searchResponse from '../../mocks/searchResponse';
import { cleanYouTubeResponse } from '../../utils/helpers';

describe('<RelatedVideos>', () => {
  const cleanSearchResponse = cleanYouTubeResponse(searchResponse.items);
  const mockResponseLength = cleanSearchResponse.length;

  afterEach(() => {
    cleanup();
  });

  it('renders', () => {
    render(
      <MemoryRouter>
        <RelatedVideos videos={cleanSearchResponse} />
      </MemoryRouter>
    );

    const title = screen.getByText('Related Videos');
    expect(title.tagName).toBe('H2');

    const links = screen.queryAllByTestId('link');
    expect(links[0].tagName).toBe('A');
    expect(links).toHaveLength(mockResponseLength);

    const videoItems = screen.queryAllByTestId('video-item');
    expect(videoItems[0].tagName).toBe('DIV');
    expect(videoItems).toHaveLength(mockResponseLength);

    const thumbnails = screen.queryAllByTestId('thumbnail');
    expect(thumbnails[0].tagName).toBe('IMG');
    expect(thumbnails).toHaveLength(mockResponseLength);

    const titles = screen.queryAllByTestId('title');
    expect(titles[0].tagName).toBe('H4');
    expect(titles).toHaveLength(mockResponseLength);
  });

  it('matches snapshot', () => {
    const { container } = render(
      <MemoryRouter>
        <RelatedVideos videos={cleanSearchResponse} />
      </MemoryRouter>
    );

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it('renders an error message when no videos prop is provided', () => {
    const { container } = render(
      <MemoryRouter>
        <RelatedVideos />
      </MemoryRouter>
    );

    expect(container.innerHTML).toMatch('No related videos were found');
  });
});
