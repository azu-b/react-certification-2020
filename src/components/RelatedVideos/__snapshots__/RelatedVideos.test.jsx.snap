// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RelatedVideos> matches snapshot 1`] = `
"<div class=\\"sc-gtssRu bMBAbX\\">
  <h2 class=\\"sc-bdnylx eGPSfR\\">Related Videos</h2>
  <div data-testid=\\"video-list\\" class=\\"sc-gtssRu sc-dlnjPT bMBAbX cJMenI\\"><a data-testid=\\"link\\" href=\\"/video/3w5hjn7sY8g\\">
      <div data-testid=\\"video-item\\" class=\\"sc-hKFyIo hztzrr\\"><img data-testid=\\"thumbnail\\" src=\\"https://i.ytimg.com/vi/3w5hjn7sY8g/default.jpg\\" class=\\"sc-eCApGN dFuKii\\">
        <h4 data-testid=\\"title\\" class=\\"sc-jSFkmK eAifiS\\">HIGH ELO Best Champions TIER List - League of Legends Patch 10.20</h4>
      </div>
    </a><a data-testid=\\"link\\" href=\\"/video/RR0JFDWuceo\\">
      <div data-testid=\\"video-item\\" class=\\"sc-hKFyIo hztzrr\\"><img data-testid=\\"thumbnail\\" src=\\"https://i.ytimg.com/vi/RR0JFDWuceo/default.jpg\\" class=\\"sc-eCApGN dFuKii\\">
        <h4 data-testid=\\"title\\" class=\\"sc-jSFkmK eAifiS\\">When LOL Players PERFECT ESCAPE... 200 IQ ESCAPES MONTAGE (League of Legends)</h4>
      </div>
    </a><a data-testid=\\"link\\" href=\\"/video/o1YfnQ5enz8\\">
      <div data-testid=\\"video-item\\" class=\\"sc-hKFyIo hztzrr\\"><img data-testid=\\"thumbnail\\" src=\\"https://i.ytimg.com/vi/o1YfnQ5enz8/default.jpg\\" class=\\"sc-eCApGN dFuKii\\">
        <h4 data-testid=\\"title\\" class=\\"sc-jSFkmK eAifiS\\">THIS ITEM MAKE SHACO PERMINENTLY INVISIBLE (NO COUNTER) - BunnyFuFuu | League Of Legends | Season 11</h4>
      </div>
    </a><a data-testid=\\"link\\" href=\\"/video/aR-KAldshAE\\">
      <div data-testid=\\"video-item\\" class=\\"sc-hKFyIo hztzrr\\"><img data-testid=\\"thumbnail\\" src=\\"https://i.ytimg.com/vi/aR-KAldshAE/default.jpg\\" class=\\"sc-eCApGN dFuKii\\">
        <h4 data-testid=\\"title\\" class=\\"sc-jSFkmK eAifiS\\">Warriors | Season 2020 Cinematic - League of Legends (ft. 2WEI and Edda Hayes)</h4>
      </div>
    </a><a data-testid=\\"link\\" href=\\"/video/pngf7xqrsDU\\">
      <div data-testid=\\"video-item\\" class=\\"sc-hKFyIo hztzrr\\"><img data-testid=\\"thumbnail\\" src=\\"https://i.ytimg.com/vi/pngf7xqrsDU/default.jpg\\" class=\\"sc-eCApGN dFuKii\\">
        <h4 data-testid=\\"title\\" class=\\"sc-jSFkmK eAifiS\\">NEW UPDATES: Seraphine Abilities LEAKED? + Upcoming JG &amp;amp; Mid Nerfs &amp;amp; MORE - League of Legends</h4>
      </div>
    </a></div>
</div>"
`;
